File: Bayes_Gaussian_Gibbs.cpp
#include <iostream>
#include <cmath>
#include<fstream>
const int niter=10001000;
const double av_x=-0.0930181e0;
const double av_y=0.0475899e0;
const double av_xx=1.06614e0;
const double av_yy=1.28152e0;
const double av_xy=-0.504944e0;
const int nsample=100;
const int ndim=2;//don't chage this parameter.
const double step_A=0.1e0;
const int nskip=10;
/******************************************************************/
/*** Gaussian Random Number Generator with Box Muller Algorithm ***/
/******************************************************************/
int BoxMuller(double& p, double& q){
  
  double pi=2e0*asin(1e0);
  //uniform random numbers between 0 and 1
  double r = (double)rand()/RAND_MAX;
  double s = (double)rand()/RAND_MAX;
  //Gaussian random numbers, 
  //with weights proportional to e^{-p^2/2} and e^{-q^2/2}
  p=sqrt(-2e0*log(r))*sin(2e0*pi*s);
  q=sqrt(-2e0*log(r))*cos(2e0*pi*s);
  
  return 0;
}
/*********************************/
/*** Calculation of the action ***/
/*********************************/
double calc_action(const double A[ndim][ndim],const double mu[ndim]){
  
  double action=
    A[0][0]*((mu[0]-av_x)*(mu[0]-av_x)+av_xx-av_x*av_x)
    +A[1][1]*((mu[1]-av_y)*(mu[1]-av_y)+av_yy-av_y*av_y)
    +A[0][1]*((mu[0]-av_x)*(mu[1]-av_y)+av_xy-av_x*av_y)*2e0
    -log(A[0][0]*A[1][1]-A[0][1]*A[1][0]);
  action=0.5e0*action*nsample; 
  //add the contribution from the prior distribution
  action=action+0.5e0*A[0][0]*A[0][0]+0.5e0*A[0][0]*A[1][1]+A[0][1]*A[0][1]+0.5e0*mu[0]*mu[0]+0.5e0*mu[1]*mu[1];
  
  return action;
}
int main()
{
  double mu[ndim];
  double A[ndim][ndim];
  double dx;
  int naccept=0;
  srand((unsigned)time(NULL));
  /*********************************/
  /* Set the initial configuration */
  /*********************************/
  //NEVER choose a singular initial condition.
  A[0][0]=1e0;A[1][1]=1e0;
  A[0][1]=0e0;
  A[1][0]=A[0][1];
  mu[0]=0e0;mu[1]=0e0;
  /*****************/
  /*** Main part ***/
  /*****************/
  for(int iter=0; iter!=niter; iter++){
    /*************************/
    /* Gibbs sampling for mu */
    /*************************/
    double r1,r2;
    BoxMuller(r1,r2);
    mu[0]=r1/sqrt(1e0+nsample*A[0][0]);
    mu[0]=mu[0]-nsample/(1e0+nsample*A[0][0])*(A[0][1]*(mu[1]-av_y)-A[0][0]*av_x);

    mu[1]=r2/sqrt(1e0+nsample*A[1][1]);
    mu[1]=mu[1]-nsample/(1e0+nsample*A[1][1])*(A[1][0]*(mu[0]-av_x)-A[1][1]*av_y);
    /********************/
    /* Metropolis for A */
    /********************/
    double backup_A[ndim][ndim];
    backup_A[0][0]=A[0][0];
    backup_A[1][1]=A[1][1];
    backup_A[0][1]=A[0][1];
    backup_A[1][0]=A[1][0];

    double action_init=calc_action(A,mu);
    dx = (double)rand()/RAND_MAX-0.5e0;
    A[0][0]=A[0][0]+dx*step_A*2e0;
    dx = (double)rand()/RAND_MAX-0.5e0;
    A[1][1]=A[1][1]+dx*step_A*2e0;
    dx = (double)rand()/RAND_MAX-0.5e0;
    A[1][0]=A[1][0]+dx*step_A*2e0;
    A[0][1]=A[0][1]+dx*step_A*2e0;
    double action_fin=calc_action(A,mu);
    /*******************/
    /* Metropolis test */
    /*******************/
    double metropolis = (double)rand()/RAND_MAX;    
    if(exp(action_init-action_fin) > metropolis){
      /* accept */
      naccept=naccept+1;
    }else{ 
      /* reject */
      A[0][0]=backup_A[0][0];
      A[1][1]=backup_A[1][1];
      A[0][1]=backup_A[0][1];
      A[1][0]=backup_A[1][0];
    }

    // output 
    if((iter+1)%nskip==0){
      std::cout << std::fixed << std::setprecision(6) 
        << A[0][0] << "  "
        << A[1][1] << "  "
        << A[0][1] << "  "
        << mu[0] << "  "
        << mu[1] << "  "
        << (double) naccept/(iter+1) << std::endl;
    }
  } 
  
  return 0;
}
  

File: Bayes_Gaussian_Gibbs.py
import numpy as np
import copy
from matplotlib import pyplot as plt
# seed of random numbers is the system time by default
# Numpy has Gaussian random number generator, so we don't use Box-Muller

NITER=100000
av_x=-0.0930181
av_y=0.0475899
av_xx=1.06614
av_yy=1.28152
av_xy=-0.504944
NSAMPLE=100
NDIM=2 #don't chage this parameter.
STEP_A=0.1
NSKIP=10
#################################
### Calculation of the action ###
#################################
def calc_action(A,mu):
    action = \
    A[0][0]*((mu[0]-av_x)*(mu[0]-av_x)+av_xx-av_x*av_x)\
    +A[1][1]*((mu[1]-av_y)*(mu[1]-av_y)+av_yy-av_y*av_y)\
    +A[0][1]*((mu[0]-av_x)*(mu[1]-av_y)+av_xy-av_x*av_y)*2.0\
    -np.log(A[0][0]*A[1][1]-A[0][1]*A[1][0])

    action=0.5*action*NSAMPLE

    #add the contribution from the prior distribution
    action += \
    0.5*A[0][0]*A[0][0] + 0.5*A[1][1]*A[1][1] + A[0,1]*A[0,1]\
    + 0.5*mu[0]*mu[0] + 0.5*mu[1]*mu[1]
    return action
  
#####################################
### Set the initial configuration ###
#####################################
#NEVER choose a singular initial condition.
A = [[1.0,0.0],
     [0.0,1.0]]
mu =[0.0,0.0]
A = np.array(A)
mu = np.array(mu)
#################
### Main part ###
#################
naccept = 0 #counter for the number of acceptance

# for plot
A00=[]
A11=[]
A01=[]
mu0=[]
mu1=[]

for iter in range(NITER):
    #############################
    ### Gibbs sampling for mu ###
    #############################
    r = np.random.randn()
    mu[0] = r/np.sqrt(1.0+NSAMPLE*A[0][0])
    mu[0] += -NSAMPLE/(1.0+NSAMPLE*A[0][0])*(A[0][1]*(mu[1]-av_y)-A[0][0]*av_x)
    r = np.random.randn()
    mu[1] = r/np.sqrt(1.0+NSAMPLE*A[1][1])
    mu[1] += -NSAMPLE/(1.0+NSAMPLE*A[1][1])*(A[1][0]*(mu[0]-av_x)-A[1][1]*av_y)
    ########################
    ### Metropolis for A ###
    ########################
    backup_A = copy.deepcopy(A)
    action_init = calc_action(A,mu)
    dx = np.random.uniform(-STEP_A, STEP_A)
    A[0][0] += dx
    dx = np.random.uniform(-STEP_A, STEP_A)
    A[1][1] += dx
    dx = np.random.uniform(-STEP_A, STEP_A)
    A[1][0] += dx
    A[0][1] += dx
    action_fin = calc_action(A,mu)
    #######################
    ### Metropolis test ###
    #######################
    metropolis = np.random.uniform(0,1)
    if(np.exp(action_init-action_fin) > metropolis):#accept
        naccept += 1
    else:#reject
        A = copy.deepcopy(backup_A)
    ###################
    ### data output ###
    ###################
    if(iter%NSKIP == 0):
        print(A[0][0],A[1][1],A[0][1],mu[0],mu[1],naccept/(iter+1))
        if( iter > 1000 ):
          A00.append(A[0][0])
          A11.append(A[1][1])
          A01.append(A[0][1])
          mu0.append(mu[0])
          mu1.append(mu[1])

############
### plot ###
############
plt.figure(figsize=(12,6))
plt.axes().set_aspect('equal')
##
ax1 = plt.subplot(2, 3, 1)
ax1.hist(A00,bins=50,density=True)
ax1.set_xlabel("$A_{11}$")
##
ax2 = plt.subplot(2, 3, 2)
ax2.hist(A11,bins=50,density=True)
ax2.set_xlabel("$A_{22}$")
##
ax3 = plt.subplot(2, 3, 3)
ax3.hist(A01,bins=50,density=True)
ax3.set_xlabel("$A_{12}=A_{21}$")
##
ax4 = plt.subplot(2, 3, 4)
ax4.hist(mu0,bins=50,density=True)
ax4.set_xlabel("$\mu_1$")
##
ax5 = plt.subplot(2, 3, 5)
ax5.hist(mu1,bins=50,density=True)
ax5.set_xlabel("$\mu_2$")
##
plt.show()



